---
title: "AMR-16S for R mini workshop in SWS6323"
author: "JM"
date: "2024-April"
output: html_document
---

Load libraries

```{r, echo=FALSE}
library(dada2)
library(ggplot2)
library(phyloseq)
library(vegan)
library(CoDaSeq)
library(dplyr)
library(tibble)
library(randomcoloR)
library(ANCOMBC)
writeLines(capture.output(sessionInfo()), "sessionInfo.txt")
```

## Step 1: Quality-filter the Illumina sequencing reads and create an Amplicon Sequence Variant (ASV) table with DADA2 - this step was completed in class on 4/12/24

DADA2 tutorial: https://benjjneb.github.io/dada2/tutorial_1_8.html

Put unjoined R1 and R2 fastq files, with adaptors and primers previously removed with cutadapt into a directory for DADA2. Here, our forward and reverse fastq filenames have format: SAMPLENAME_R1_cut.fastq.gz and SAMPLENAME_R2_cut.fastq.gz

# Be sure to update the path names on line 34 and line 82 as needed.

This chunk of code will save the ASV table as an .rds file that can be reopened later. Please note that if you repeat the DADA2 step, you will get a slightly different table every time. If you have sequencing data from multiple sequencing runs, you should run DADA2 on each sequencing run and then merge the tables as described here: https://benjjneb.github.io/dada2/bigdata_paired.html

```{r, echo=FALSE}
path <- "~/Documents/SWS6323_MiniWorkshop/cutadapt" # adjust file name as needed
list.files(path)
fnFs <- sort(list.files(path, pattern="_R1_cut.fastq.gz", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_R2_cut.fastq.gz", full.names = TRUE))
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
# Perform filtering and trimming
filt_path <- file.path(path, "filtered") 
filtFs <- file.path(filt_path, paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(filt_path, paste0(sample.names, "_R_filt.fastq.gz"))
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(150,150),
                     maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
                     compress=TRUE, multithread=TRUE)
head(out)
# Learn the Error Rates, it TAKES TIME!
errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)
plotErrors(errF, nominalQ=TRUE)
# Dereplicate the filtered fastq files
derepFs <- derepFastq(filtFs, verbose=TRUE)
derepRs <- derepFastq(filtRs, verbose=TRUE)
names(derepFs) <- sample.names
names(derepRs) <- sample.names
# Infer the sequence variants in each sample
dadaFs <- dada(derepFs, err=errF, multithread=TRUE)
dadaRs <- dada(derepRs, err=errR, multithread=TRUE)
# Inspecting the dada-class object returned by dada:
dadaFs[[1]]
# Merge the denoised forward and reverse reads:
mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE)
# Inspect the merger data.frame from the first sample
head(mergers[[1]])
# Construct sequence table
seqtab <- makeSequenceTable(mergers)
dim(seqtab)
#Remove chimeric sequences:
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
sum(seqtab.nochim)/sum(seqtab)
# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))
# Track reads through the pipeline
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(mergers, getN), rowSums(seqtab))
colnames(track) <- c("input", "filtered", "denoised", "merged", "tabled")
rownames(track) <- sample.names
head(track)
write.table(track, "dada_read_stats.txt",sep="\t",col.names=NA) # Examine and save the stats table as a summary of the quality of your sequencing data. 

# SAVE THIS FILE SO YOU DON'T HAVE TO REPEAT ALL OF THE ABOVE STEPS, adjust name as needed
saveRDS(seqtab.nochim, "~/Documents/SWS6323_MiniWorkshop/seqtab.rds")
```

## Step 2: Assign taxonomy in DADA2 - this step completed in class 4/19/24

Note: my strategy for naming files is to use the same file names across projects, but organize the output into directories that indicate the project name. This saves me time when re-using code. I have indicated where file names should be updated as needed.

Make sure the taxonomy reference database is in your working directory. Keep the database file gzipped. Adjust path name below. This step is very time consuming.

When taxonomy assignment is complete, we will use base R and phyloseq to clean up the taxonomy table. First, we will replace NAs and empty cells with the lowest taxonomy classification available. Second, we will use phyloseq to remove reads that are classified as Eukaryotes or unclassified at the domain level (ie, we are keeping only Bacteria and Archaea because that is what our primers target).

After generating the taxonomy table, we will create a phyloseq object. This "object" is the intersection of 3 data matrices: the ASV table generated with DADA2, the taxonomy table we generate with DADA2, and a table of metadata that you provide at this point. It is critical that the sample names in your metadata file exactly match the sample names on your files that went into DADA2. If the names do not match, this script will still generate a phyloseq object, but it will be missing some of the samples. Check that the number of samples is correct when you first generate the phyloseq object.

```{r, echo=FALSE}
# RELOAD THE SAVED INFO FROM HERE (if you have closed the project):
seqtab.nochim <- readRDS("~/Documents/SWS6323_MiniWorkshop/seqtab.rds")
taxa <- assignTaxonomy(seqtab.nochim, "~/Documents/SWS6323_MiniWorkshop/silva_nr99_v138.1_train_set.fa.gz", multithread=TRUE) # PC users: set multithread=FALSE

# FIX the NAs in the taxa table - this chunk of code (lines 101-108) is a custom script written by Dr. Meyer, you will not find it in tutorials for DADA2 or phyloseq
taxon <- as.data.frame(taxa,stringsAsFactors=FALSE)
taxon$Phylum[is.na(taxon$Phylum)] <- taxon$Kingdom[is.na(taxon$Phylum)]
taxon$Class[is.na(taxon$Class)] <- taxon$Phylum[is.na(taxon$Class)]
taxon$Order[is.na(taxon$Order)] <- taxon$Class[is.na(taxon$Order)]
taxon$Family[is.na(taxon$Family)] <- taxon$Order[is.na(taxon$Family)]
taxon$Genus[is.na(taxon$Genus)] <- taxon$Family[is.na(taxon$Genus)]
write.table(taxon,"silva_taxa_table.txt",sep="\t",col.names=NA)
write.table(seqtab.nochim, "silva_otu_table.txt",sep="\t",col.names=NA)

# Create a phyloseq object from the ASV table and the taxonomy table generated with DADA2, along with the sample metadata.
otu <- read.table("silva_otu_table.txt",sep="\t",header=TRUE, row.names=1) # reading in this data can take time - this file took 5 mins to load
taxon <- read.table("silva_taxa_table.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps # check that all of your samples are included: 6756 taxa and 52 samples

# use phyloseq to remove chloroplasts and mitochondria and Eukaryota
get_taxa_unique(ps, "Family") # 481
get_taxa_unique(ps, "Order") # 284
get_taxa_unique(ps, "Kingdom") # 3
ps <- subset_taxa(ps, Family !="Mitochondria")
ps <- subset_taxa(ps, Order !="Chloroplast")
ps <- subset_taxa(ps, Kingdom !="Eukaryota")
ps <- subset_taxa(ps, Kingdom !="NA")
get_taxa_unique(ps, "Family") # 478
get_taxa_unique(ps, "Order") # 282
get_taxa_unique(ps, "Kingdom") # 2
ps  #6141  taxa and 52 samples

# Now export cleaned otu and taxa tables from phyloseq for future reference
otu = as(otu_table(ps), "matrix")
taxon = as(tax_table(ps), "matrix")
metadata = as(sample_data(ps), "matrix")
write.table(otu,"silva_nochloronomito_otu_table.txt",sep="\t",col.names=NA)
write.table(taxon,"silva_nochloronomito_taxa_table.txt",sep="\t",col.names=NA)

# export ASV table as relative abundance
ps_ra<-transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
otu_ra = as(otu_table(ps_ra), "matrix")
write.table(otu_ra,"silva_nochloronomito_otu_table_RA.txt",sep="\t",col.names=NA)

```

## Step 3: Explore the data with phyloseq - this step completed in class 4/19/24

In this step we will also use phyloseq to remove sequences that are not classified as Bacteria or Archaea and to remove sequences classified as chloroplasts or mitochondria. At the end of this step we will save the tables for future use, including a table with relative abundances instead of the raw count data.


```{r, echo=FALSE}
# load in data and create phyloseq object
otu <- read.table("silva_nochloronomito_otu_table.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("silva_nochloronomito_taxa_table.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps # 6141 taxa and 52 samples
psnb = subset_samples(ps, colony != "blank") #remove sample blanks
psnb # 6141 taxa and 50 samples
# remove disease samples to compare before and after amox treatment
ps1 = subset_samples(psnb, type != "disease") 
ps1 # 6141 taxa and 35 samples
otu = as(otu_table(ps1), "matrix")
taxon = as(tax_table(ps1), "matrix")
metadata = as(sample_data(ps1), "matrix")
write.table(otu,"silva_nochloronomito_otu_table_ps1.txt",sep="\t",col.names=NA)
write.table(taxon,"silva_nochloronomito_taxa_table_ps1.txt",sep="\t",col.names=NA)
write.table(metadata,"metadata_ps1.txt",sep="\t",col.names=NA)

ps5<-filter_taxa(ps1, function(x) mean(x) >5, TRUE)
ntaxa(ps5) #112
ps10<-filter_taxa(ps1, function(x) mean(x) >10, TRUE)
ntaxa(ps10) #78
get_taxa_unique(ps1, "Genus") #886
get_taxa_unique(ps5, "Genus") #61
get_taxa_unique(ps10, "Genus") #46

# Now export filtered otu and taxa tables from phyloseq for future reference
otu_ps5 = as(otu_table(ps5), "matrix")
taxon_ps5 = as(tax_table(ps5), "matrix")
metadata = as(sample_data(ps5), "matrix")
write.table(otu_ps5,"silva_nochloronomito_otu_table_ps5.txt",sep="\t",col.names=NA)
write.table(taxon_ps5,"silva_nochloronomito_taxa_table_ps5.txt",sep="\t",col.names=NA)
write.table(metadata,"metadata_ps5.txt",sep="\t",col.names=NA) 

#relative abundance
ps5_ra<-transform_sample_counts(ps5, function(OTU) OTU/sum(OTU))
otu_ps5_ra = as(otu_table(ps5_ra), "matrix")
write.table(otu_ps5_ra,"silva_nochloronomito_otu_table_ps5_RA.txt",sep="\t",col.names=NA)

```

## Step 4: Examine ordinations of the data - this step completed in class 4/19/24
# Pipeline in this block of code: ASV table -> CLR transformation -> Aitchison distance -> PCA ordination

This block of code will perform center-log-ratio transformation on ASVs and calculate Aitchison Distance and principal components. This is the recommended pipeline for compositional datasets. You can begin this chunk of code with an empty R environment - in other words, you can start here without repeating any of the above steps.

The PERMANOVA test to accompany the PCA ordination will be completed in class 4/22/24.

```{r, echo=FALSE}
# load in data and create phyloseq object
otu <- read.table("silva_nochloronomito_otu_table_ps5.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("silva_nochloronomito_taxa_table_ps5.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata_ps5.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps5 <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps5 #112 taxa and 35 samples

# First, replace 0 values with an estimate (because normalization is taking log, can't have 0)
# Also transposing here, need samples as rows
d.czm <- cmultRepl(t(otu), method="CZM", label=0, z.warning=1)
# Perform the center-log-ratio (CLR) transformation 
d.clr <- codaSeq.clr(d.czm)
# transpose matrix of CLR transformed data for ordination and dendrogram
E.clr <- t(d.clr)
# plot compositional PCA biplot (perform a singular value decomposition)
d.pcx <- prcomp(E.clr)
# calculate percent variance explained for the axis labels
pc1 <- round(d.pcx$sdev[1]^2/sum(d.pcx$sdev^2),2)
pc2 <- round(d.pcx$sdev[2]^2/sum(d.pcx$sdev^2),2)
xlab <- paste("PC1: ", pc1, sep="")
ylab <- paste("PC2: ", pc2, sep="")
#biplot(d.pcx, cex=c(0.6,0.4), var.axes=F,scale=1, xlab=xlab, ylab=ylab)
summary(d.pcx)
str(d.pcx)
screeplot(d.pcx)
# Make a pretty PCA plot with ggplot
df_out <- as.data.frame(d.pcx$x)
theme_set(theme_bw()+theme(panel.grid.major = element_blank(),panel.grid.minor = element_blank()))
cols<-c("before"="#6CD5D9","after"="#107A86")
samples$type<-factor(samples$type, levels=c("before","after"))
pdf("PCA.pdf",bg ="white",width=8.5)
p<-ggplot(df_out,aes(x=PC1,y=PC2,fill=samples$type,shape=samples$fraction))
p<-p+geom_point(size=3)+
  theme(axis.title = element_text(size=14))+
  theme(axis.text=element_text(size=12))+
  theme(legend.title = element_text(size=14))+
  theme(legend.text = element_text(size=12))+
  scale_shape_manual(values=c(21,22))+
  scale_fill_manual(values=cols)+
  guides(fill = guide_legend(override.aes=list(shape=21)))+
  facet_grid(~samples$fraction)
p + labs(x=xlab, y=ylab, fill="Type",shape="Fraction") +coord_fixed()
dev.off()


# For comparison, PCA without faceting:
pdf("PCA_nofacet.pdf",bg ="white",width=8.5)
p<-ggplot(df_out,aes(x=PC1,y=PC2,fill=samples$type,shape=samples$fraction))
p<-p+geom_point(size=3)+
  theme(axis.title = element_text(size=14))+
  theme(axis.text=element_text(size=12))+
  theme(legend.title = element_text(size=14))+
  theme(legend.text = element_text(size=12))+
  scale_shape_manual(values=c(21,22))+
  scale_fill_manual(values=cols)+
  guides(fill = guide_legend(override.aes=list(shape=21)))
p + labs(x=xlab, y=ylab, fill="Type",shape="Fraction") +coord_fixed()
dev.off()

####### Use phyloseq/vegan to perform PERMANOVA
# set metadata as factors
type<-as.character(samples$type)
frac<-as.character(samples$fraction)

# permanova between groups using Aitchison distance
dist.clr <- dist(E.clr)
perm_atch <- adonis2(dist.clr~type*frac,as(sample_data(ps5),"data.frame"))
print(perm_atch) # copy the output on the console into a text file for future reference.
```

## Step 4: Examine ordinations of the data - this step completed in class 4/19/24
# Pipeline in this block of code: ASV table -> relative abundance -> Bray-Curtis distance -> nMDS ordination

For comparison to the ordination generated above designed for compositional datasets, let's see what an ordination using a more traditional workflow looks like - specifically an nMDS ordination from the Bray-Curtis dissimilarity of relative abundance data. Again, you can begin this chunk of code with an empty R environment - in other words, you can start here without repeating any of the above steps.

```{r, echo=FALSE}
# load in data and create phyloseq object
otu <- read.table("silva_nochloronomito_otu_table_ps5.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("silva_nochloronomito_taxa_table_ps5.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata_ps5.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps5 <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps5 #112 taxa and 35 samples
ps5_ra<-transform_sample_counts(ps5, function(OTU) OTU/sum(OTU)) # convert ASV counts to relative abundance

ord <- ordinate(ps5_ra, "NMDS","bray") # this line calculates the Bray-Curtis distance and calculates the ordination
# set preferences for a pretty plot
cols<-c("before"="#6CD5D9","after"="#107A86")
samples$type<-factor(samples$type, levels=c("before","after"))
# now plot the ordination using ggplot2 through phyloseq wrapper
pdf("nMDS.pdf",bg ="white",width=8.5, height=4)
p1 <- plot_ordination(ps5_ra, ord, color="type", shape="fraction")
p1 <- p1 + facet_grid(~fraction)+
  geom_point(size=3)+
  theme(axis.title = element_text(size=14))+
  theme(axis.text=element_text(size=12))+
  theme(legend.title = element_text(size=14))+
  theme(legend.text = element_text(size=12))+
  scale_color_manual(values=cols)+
  facet_grid(~samples$fraction)
p1 +coord_fixed()
dev.off()


####### Use phyloseq/vegan to perform PERMANOVA
#calculate the Bray Curtis distance
bc <- distance(ps5_ra, method="bray", type="samples")

# set metadata as factors
type<-as.character(samples$type)
frac<-as.character(samples$fraction)

# perform PERMANOVA between groups using Bray Curtis ditance
perm_bc <- adonis2(bc~type*frac,as(sample_data(ps5_ra),"data.frame"))
print(perm_bc) # copy the output on the console into a text file for future reference.

```


## Create stacked bar charts from relative abundance data

You can begin this chunk of code with an empty R environment - in other words, you can start here without repeating any of the above steps.


```{r, echo=FALSE}
# load in data and create phyloseq object
otu <- read.table("silva_nochloronomito_otu_table_ps5.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("silva_nochloronomito_taxa_table_ps5.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata_ps5.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps5 <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps5 #112 taxa and 35 samples

# Convert ASV counts to relative abundance for plotting
ps5_ra<-transform_sample_counts(ps5, function(OTU) OTU/sum(OTU))
# In the next four lines I am seeing how many colors I would need in my legend to accommodate each taxonomic level
get_taxa_unique(ps5_ra, "Class") #19
get_taxa_unique(ps5_ra, "Order") #37
get_taxa_unique(ps5_ra, "Family") #52
get_taxa_unique(ps5_ra, "Genus") #61

# manually setting the order of my metadata because the default in R is alphabetical
sample_data(ps5_ra)$type<-factor(sample_data(ps5_ra)$type, levels=c("before","after"))

n <- 19
# after plotting, you can re-run the next line to create a different selection of colors
palette <- distinctColorPalette(n)
print(palette)

pdf("barchart_Class.pdf",width=11)
p1=plot_bar(ps5_ra, "colony" ,fill="Class")+
  facet_grid(fraction~type,scales="free",space="free")+
  geom_bar(aes(fill=Class), stat="identity",position="stack")+
  theme_bw()+
  theme(strip.text=element_text(face="bold", size=12))+
  theme(axis.text.x=element_text(size=12))+
  theme(axis.text.y=element_text(size=12))+
  scale_fill_manual(values=palette)+
  #theme(axis.title.x = element_blank())+
  theme(legend.position = "bottom")
p1
dev.off()



n <- 37
palette <- distinctColorPalette(n)

pdf("barchart_Order.pdf",width=11)
p1=plot_bar(ps5_ra, "colony" ,fill="Order")+
  facet_grid(fraction~type,scales="free",space="free")+
  geom_bar(aes(fill=Order), stat="identity",position="stack")+
  theme_bw()+
  theme(strip.text=element_text(face="bold", size=12))+
  theme(axis.text.x=element_text(size=12))+
  theme(axis.text.y=element_text(size=12))+
  scale_fill_manual(values=palette)+
  #theme(axis.title.x = element_blank())+
  theme(legend.position = "bottom")
p1
dev.off()


n <- 52
palette <- distinctColorPalette(n)

pdf("barchart_Family.pdf",width=11)
p1=plot_bar(ps5_ra, "colony" ,fill="Family")+
  facet_grid(fraction~type,scales="free",space="free")+
  geom_bar(aes(fill=Family), stat="identity",position="stack")+
  theme_bw()+
  theme(strip.text=element_text(face="bold", size=12))+
  theme(axis.text.x=element_text(size=12))+
  theme(axis.text.y=element_text(size=12))+
  scale_fill_manual(values=palette)+
  #theme(axis.title.x = element_blank())+
  theme(legend.position = "bottom")
p1
dev.off()

n <- 61
palette <- distinctColorPalette(n)

pdf("barchart_Genus.pdf",width=11)
p1=plot_bar(ps5_ra, "colony" ,fill="Genus")+
  facet_grid(fraction~type,scales="free",space="free")+
  geom_bar(aes(fill=Genus), stat="identity",position="stack")+
  theme_bw()+
  theme(strip.text=element_text(face="bold", size=12))+
  theme(axis.text.x=element_text(size=12))+
  theme(axis.text.y=element_text(size=12))+
  scale_fill_manual(values=palette)+
  #theme(axis.title.x = element_blank())+
  theme(legend.position = "bottom")
p1
dev.off()


```

## Differential abundance of taxa across treatments using ANCOM-BC2

There are many different programs you can choose to examine the differential abundance of taxa among your groups of samples. Nearing et al 2022 (https://www.nature.com/articles/s41467-022-28034-z) tested the performance of several commonly used tools and found that ANCOM-BC2 was one of the best performers and it is designed specifically for dealing with the compositional datasets of microbiomes.

Here is the online tutorial for ANCOM-BC2:
https://bioconductor.org/packages/release/bioc/vignettes/ANCOMBC/inst/doc/ANCOMBC2.html

This tutorial is not as easy to use as the DADA2 tutorial. There are many options for customization, but this tutorial does not list out what all the options mean, as I frequently see with programs you would run on the command line.

ANCOM-BC2 will perform its own normalization - it does not use the CLR transformation we used for the PCA ordination - therefore you will need to start with your full ASV table which will take some time to load.

This chunk of code can be run with a clean environment, without repeating any of the above steps The input for ANCOM-BC2 used here is a phyloseq object, so we start by reading in the ASV table, the taxonomy table, and the metadata and create that phyloseq object.  Notice that you will be using the dataset that does not have diseased or blank samples and *does not have low abundance ASVs removed*.

After you have your phyloseq object, you will want to understand what you are asking ANCOM-BC2 to test. ANCOM-BC2 results will be expressed as a relative relationship - it will test variables in the order that they are input. The default in R is alphabetical order, but I want to compare taxa "after" relative to "before", so I need to manually state the order of input so that "before" is input first.


```{r, echo=FALSE}
# Start by creating phyloseq object using full dataset (low abundance ASVs are not removed).

otu <- read.table("silva_nochloronomito_otu_table_ps1.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("silva_nochloronomito_taxa_table_ps1.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata_ps1.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps1 <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps1 #6141 taxa and 35 samples

# ANCOM-BC2 results will be expressed as a relative relationship - it will test variables in the order that they are input. The default in R is alphabetical order, but I want to compare taxa "after" relative to "before", so I need to manually state the order of input.
sample_data(ps1)$type<-factor(sample_data(ps1)$type,levels=c("before","after"))
# You can verify the change by checking:
levels(sample_data(ps1)$type)

# Using ANCOM-BC2 to detect differentially abundant taxa between treatment types.
set.seed(123)
output = ancombc2(data = ps1, assay_name = "counts", tax_level = "Genus",
                  fix_formula = "fraction + type", rand_formula = NULL,
                  p_adj_method = "holm", pseudo = 0, pseudo_sens = TRUE,
                  prv_cut = 0.10, lib_cut = 1000, s0_perc = 0.05,
                  group = "type", struc_zero = FALSE, neg_lb = FALSE,
                  alpha = 0.05, n_cl = 2, verbose = TRUE,
                  global = FALSE, pairwise = FALSE, 
                  dunnet = FALSE, trend = FALSE,
                  iter_control = list(tol = 1e-5, max_iter = 20, 
                                      verbose = FALSE),
                  em_control = list(tol = 1e-5, max_iter = 100),
                  lme_control = NULL, mdfdr_control = NULL, 
                  trend_control = NULL)

res_prim = output$res
output$res
write.table(res_prim,file="ancomBC-results-BH-genus-by-type.txt",sep="\t") # getting "NA" for some taxa - need to find out why it is doing this

```




